virtualizarr/readers/kerchunk.py

```diff
@@ -1,310 +1,69 @@
-def _automatically_determine_filetype(
-    *,
-    filepath: str,
-    reader_options: Optional[dict[str, Any]] = {},
-) -> FileType:
-    if Path(filepath).suffix == ".zarr":
-        # TODO we could imagine opening an existing zarr store, concatenating it, and writing a new virtual one...
-        raise NotImplementedError()
-
-    # Read magic bytes from local or remote file
-    fpath = _FsspecFSFromFilepath(
-        filepath=filepath, reader_options=reader_options
-    ).open_file()
-    magic_bytes = fpath.read(8)
-    fpath.close()
-
-    if magic_bytes.startswith(b"CDF"):
-        filetype = FileType.netcdf3
-    elif magic_bytes.startswith(b"\x0e\x03\x13\x01"):
-        raise NotImplementedError("HDF4 formatted files not supported")
-    elif magic_bytes.startswith(b"\x89HDF"):
-        filetype = FileType.hdf5
-    elif magic_bytes.startswith(b"GRIB"):
-        filetype = FileType.grib
-    elif magic_bytes.startswith(b"II*"):
-        filetype = FileType.tiff
-    elif magic_bytes.startswith(b"SIMPLE"):
-        filetype = FileType.fits
-    else:
-        raise NotImplementedError(
-            f"Unrecognised file based on header bytes: {magic_bytes}"
-        )
-
-    return filetype
-
-
-def read_kerchunk_references_from_file(
-    filepath: str,
-    filetype: FileType | None,
-    group: str | None,
-    reader_options: Optional[dict[str, Any]] = None,
-) -> KerchunkStoreRefs:
-    """
-    Read a single legacy file and return kerchunk references to its contents.
-
-    Parameters
-    ----------
-    filepath : str, default: None
-        File path to open as a set of virtualized zarr arrays.
-    filetype : FileType, default: None
-        Type of file to be opened. Used to determine which kerchunk file format backend to use.
-        If not provided will attempt to automatically infer the correct filetype from the the filepath's extension.
-    group : str, default is None
-        Path to the HDF5/netCDF4 group in the given file to open. Given as a str, supported by filetypes “netcdf4” and “hdf5”.
-        Dict passed into Kerchunk file readers. Note: Each Kerchunk file reader has distinct arguments,
-        so ensure reader_options match selected Kerchunk reader arguments.
-    """
-
-    if reader_options is None:
-        reader_options = {}
-
-    if filetype is None:
-        filetype = _automatically_determine_filetype(
-            filepath=filepath, reader_options=reader_options
+from virtualizarr.utils import _FsspecFSFromFilepath, check_for_collisions
+
+
+class KerchunkVirtualBackend(VirtualBackend):
+    @staticmethod
+    def open_virtual_dataset(
+        filepath: str,
+        group: str | None = None,
+        drop_variables: Iterable[str] | None = None,
+        loadable_variables: Iterable[str] | None = None,
+        decode_times: bool | None = None,
+        indexes: Mapping[str, Index] | None = None,
+        reader_options: Optional[dict] = None,
+    ) -> Dataset:
```

Comment:
Another nice thing about this approach is that it makes it really clear which readers are missing which features. For example `drop_variables` did not exist for the kerchunk reader (I've now added a simple implementation of it).